File: src/main/java/de/wittenbude/exportify/models/Album.java
package de.wittenbude.exportify.models;

import com.neovisionaries.i18n.CountryCode;
import de.wittenbude.exportify.models.embeds.Copyright;
import de.wittenbude.exportify.models.embeds.ExternalIDs;
import de.wittenbude.exportify.models.embeds.Image;
import de.wittenbude.exportify.models.embeds.ReleaseDatePrecision;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.Id;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.proxy.HibernateProxy;
import org.hibernate.type.SqlTypes;

import java.time.Instant;
import java.util.*;

@Getter
@Setter
@Accessors(chain = true)
@Entity
public class Album {

    @Id
    @GeneratedValue
    private UUID id;

    @CreationTimestamp
    private Instant versionTimestamp;

    private String albumType;
    private Integer totalTracks;
    private String href;
    private String spotifyID;
    private String name;
    private String releaseDate;
    private ReleaseDatePrecision releaseDatePrecision;
    private String restrictions;
    private String spotifyObjectType;
    private String uri;
    private ExternalIDs externalIDs;
    private String label;
    private Integer popularity;

    @JdbcTypeCode(SqlTypes.JSON)
    private Collection<Copyright> copyrights;

    @JdbcTypeCode(SqlTypes.JSON)
    private Collection<Image> images;

    @JdbcTypeCode(SqlTypes.ARRAY)
    private Collection<CountryCode> availableMarkets;

    @JdbcTypeCode(SqlTypes.JSON)
    private Map<String, String> externalUrls;

    @JdbcTypeCode(SqlTypes.ARRAY)
    private Collection<String> genres;

    @JdbcTypeCode(SqlTypes.ARRAY)
    private List<String> spotifyArtistIDs;

    @JdbcTypeCode(SqlTypes.ARRAY)
    private List<String> spotifyTrackIDs;

    @Override
    public final boolean equals(Object o) {
        if (this == o) return true;
        if (o == null) return false;
        Class<?> oEffectiveClass = o instanceof HibernateProxy ? ((HibernateProxy) o).getHibernateLazyInitializer().getPersistentClass() : o.getClass();
        Class<?> thisEffectiveClass = this instanceof HibernateProxy ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass() : this.getClass();
        if (thisEffectiveClass != oEffectiveClass) return false;
        Album album = (Album) o;
        return getId() != null && Objects.equals(getId(), album.getId());
    }

    @Override
    public final int hashCode() {
        return this instanceof HibernateProxy ? ((HibernateProxy) this).getHibernateLazyInitializer().getPersistentClass().hashCode() : getClass().hashCode();
    }
}

File: src/main/kotlin/de/wittenbude/exportify/models/Album.kt
@file:Suppress("PLUGIN_IS_NOT_ENABLED")

package de.wittenbude.exportify.models

import com.neovisionaries.i18n.CountryCode
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import se.michaelthelin.spotify.enums.AlbumType
import se.michaelthelin.spotify.enums.CopyrightType
import se.michaelthelin.spotify.enums.ModelObjectType
import se.michaelthelin.spotify.model_objects.specification.SavedAlbum
import se.michaelthelin.spotify.model_objects.specification.Album as ApiAlbum

@Serializable
data class Album(
    var addedAt: String? = null,
    val albumType: AlbumType? = null,
    val artists: List<String>? = null,
    val availableMarkets: List<CountryCode>? = null,
    val copyrights: Map<CopyrightType, String>? = null,
    val externalIds: Map<String, String>? = null,
    val externalUrls: Map<String, String>? = null,
    val genres: List<String>? = null,
    val href: String? = null,
    @SerialName("_id") val id: String,
    val images: List<String>? = null,
    val label: String? = null,
    val name: String,
    val popularity: Int? = null,
    val releaseDate: String? = null,
    val releaseDatePrecision: String? = null,
    val tracks: List<String>? = null,
    val type: ModelObjectType? = null,
    val uri: String? = null
)

fun SavedAlbum.convert(tracks: List<Track>?, artists: List<String>?) : Album {
    val album = this.album.convert(tracks, artists)
    album.addedAt = this.addedAt?.toString()
    return album
}

fun ApiAlbum.convert(tracks: List<Track>?, artists: List<String>?): Album {
    return Album(
        albumType = albumType,
        artists = artists,
        availableMarkets = availableMarkets?.asList(),
        copyrights = copyrights?.associate { it.type to it.text },
        externalIds = externalIds?.externalIds,
        externalUrls = externalUrls?.externalUrls,
        genres = genres?.toList(),
        href = href,
        id = id,
        images = images?.map { it.url },
        label = label,
        name = name,
        popularity = popularity,
        releaseDate = releaseDate,
        releaseDatePrecision = releaseDatePrecision?.getPrecision(),
        tracks = tracks?.map { it.id },
        type = type,
        uri = uri,
    )
}